README.md
# Adzerk Android SDK

An Android SDK for the Adzerk Native and UserDB APIs

## Download

Grab via gradle

```gradle
implementation 'com.adzerk:sdk:2.3.+'
```

The SDK uses Java 8 and supports a minimum Android 5.1 (API level 22).

To target a client application with minimum SDK <26, refer to the [Android Java 8 support documentation](https://developer.android.com/studio/write/java8-support).

## Usage

Create an instance of the AdzerkSdk for your `networkId`:

```kotlin
AdzerkSdk sdk = new AdzerkSdk.Builder().networkId(23L).build();
```

Create an instance of the AdzerkSdk and also provide a hostname:

```kotlin
AdzerkSdk sdk = new AdzerkSdk.Builder().networkId(23L).hostname("custom.host.com").build();
```

Build a request for placements

```kotlin
Request request = new Request.Builder()
      .addPlacement(new Placement("div1", <site_id>, <ad_types...>))
      .addPlacement(new Placement("div2", <site_id>, <ad_types...>))
      .addPlacement(new Placement("div3", <site_id>, <ad_types...>))
      .build();
```

Submit the request and get a callback for the response

```kotlin
sdk.requestPlacement(request, listener);
```

Get content and decisions from the response

```kotlin
@Override
public void success(Response response) {
    Decision firstDecision = response.getDecisions("div1").get(0);
    // ...
}
```

@since SDK v2.2.0+

## Recording Impression & Clicks

Use with the fetch ad example above.

```kotlin
// Impression pixel; fire when user sees the ad
String impUrl = decision.getImpressionUrl();
sdk.firePixel(impUrl, listener);
```

```kotlin
// Click pixel; fire when user clicks on the ad
String clickUrl = decision.getClickUrl();
sdk.firePixel(clickUrl, listener);
```

Modifying the revenue
```kotlin
// Click pixel; fire when user clicks on the ad and modify the click revenue
//   OVERRIDE: replaces the revenue value of the click/event 
//   ADDITIONAL: adds the specified value to the original revenue value of the click/event
String clickUrl = decision.getClickUrl();
sdk.firePixel(clickUrl, revenue, RevenueModifierType.OVERRIDE, null, listener);
```

Setting the gross merchandise value for the event. 
```kotlin
// Click pixel; fire when user clicks on the ad and modify the click revenue
//   OVERRIDE: replaces the revenue value of the click/event 
//   ADDITIONAL: adds the specified value to the original revenue value of the click/event
String clickUrl = decision.getClickUrl();
Float grossMerchandiseValue = 1.5f;
sdk.firePixel(clickUrl, revenue, RevenueModifierType.ADDITIONAL, grossMerchandiseValue, listener);
```

Handle the fire pixel response
```kotlin
// status: HTTP status code
// location: click target URL
@Override
public void success(FirePixelResponse response) {
   Log.d(TAG, "Fired! " +
      "status: " + response.getStatusCode() + " " +
      "location: " + response.getLocation());
}
```

## Additional Options
Additional optional parameters supported by the API may be specified via the Builder on the Request or Placement.

Example: to specify eCPM partitions for a Placement:

```kotlin
String[] ecpmPartitions = new String[]{"main", "detail", "footer"};
Request request = new Request.Builder()
    .addPlacement(new Placement("div1", 1133898L, 163)
        .setCount(ecpmPartitions.length)
        .addAdditionalOption("ecpmPartitions", ecpmPartitions))
    .build();
```

@since SDK v2.1.0+

## Distance Targeting

Distance Targeting is supported via Additional Options on the request.

```kotlin
val request = new Request.Builder()
    .addPlacement(Placement("div1", 1133898, (163)))
    .addAdditionalOption("intendedLatitude", 35.91868)
    .addAdditionalOption("intendedLongitude", -78.96001)
    .addAdditionalOption("radius", 50)
    .build()
```

## Multi-Winner Placements
A multi-winner placement returns multiple selections inside a single placement object.

To request multiple winners, specify a count for the Placement object. This requests the maximum number of winners (selections) that can be included in the placement.

```kotlin
Request request = new Request.Builder()
   .addPlacement(new Placement("div1", <site_id>, <ad_types...>).setCount(3))
   .build();
```

@since SDK v2.0.0+

## GDPR Consent

Consent preferences can be specified when building a request. For example, to set GDPR consent for tracking in the European Union (this defaults to false):

```kotlin
Request request = new Request.Builder(placements)
    .setUser(new User(key))
    .setConsent(new Consent(true))
    .build();
```

@since SDK v0.4.0+

## Building
Use gradlew to build library archive

```
./gradlew assemble
```
output: `sdk/build/outputs/aar`

## Testing
Run unit tests

```
./gradlew test
```
Specify `--rerun-tasks` to rerun up-to-date tests. To view test reports see:

output: `sdk/build/reports/tests`

## Documentation
Generate the SDK documentation

```
./gradlew javadoc
```

output: `sdk/build/docs/javadoc`

## Publishing
Update the version string properties in `sdk/build.gradle`

Either get a copy of the Kevel Engineering signing key or generate a new signing key using the instructions here:

https://blog.sonatype.com/2010/01/how-to-generate-pgp-signatures-with-maven/

Update `~/.gradle/gradle.properties` with your signing information.

```
./gradlew uploadArchive
```

Login in to Sonatype OSS, locate the staging package, close, and release it.


CHANGELOG.md
# CHANGELOG

## 2.3.0 (09/02/2022)
- update: Android SDK update for API 32 (resolves: #98)

## 2.2.1 (11/10/2021)
- new: add support for gross merchandise value in pixel fire events

## 2.2.0 (08/05/2021)
- new: add general pixel firing support #91
- new: support advertiserId in response #93

## 2.1.0 (11/22/2020)
- new: add support to specify additional optional parameters at the Request and Placement levels #75
- new: adds `matchedPoints` to DecisionResponse to support GeoDistance Targeting  #78

## 2.0.0 (10/28/2020)
- new: add support for requesting multi-winner Placements by specifying a count on Placement for ad request #74
   - breaking: `DecisionResponse getDecision(name)` has been replaced by `getDecisions(name)` which returns `List<Decision>`
   - breaking: `DecisionResponse getDecisions()` now returns a map from Placement name to decision list `Map<String, List<Decision>>`
- new: add support for `botFiltering` option #76
- fix: remote outdated `time` field from ad Request #77

## 1.0.1 (08/27/2020)
- enable minSDK 22

## 1.0.0 (08/12/2020)

* add support for Adzerk e-dash domains #68
* add Adzerk SDK version header to requests #67
* add support of custom hostname
* default placements to top-level networkId

## 0.5.0 (03/22/2020)

* Android SDK updates for API 29
* update project documentation
* add console test logging

## 0.4.0 (05/19/18)

* Support GDPR params & endpoints #64

## 0.3.0 (10/22/17)

* network library updated: Retrofit2 (v2.3.0)
* updated for SDK 26

## 0.2.0 (12/12/16)

* Fixed [#51](https://github.com/adzerk/adzerk-android-sdk/issues/51): properly parse event IDs and URLs from Native Ads API responses.

## 0.1.1 (8/7/15)

...

## 0.1.0 (8/4/15)


sdk/src/main/java/com/adzerk/android/sdk/rest/Event.java
package com.adzerk.android.sdk.rest;

/**
 * Event information included in DecisionResponse to an ad requestPlacement. Once you call an Event URL, Adzerk will track that
 * event for the creative/campaign/advertiser etc. in Adzerk reporting.
 * <p/>
 * Events in Adzerk are data about a way a user interacts with an ad. Standard events included with all Adzerk
 * accounts are clicks, impressions, and conversions. Custom events are defined by the user of the account.
 */
public class Event {

    public static final int ID_VIEW_CONVERSION = 1;
    public static final int ID_CLICK_CONVERSION = 2;
    public static final int ID_SERVER_CONVERSION = 3;
    public static final int ID_UPVOTE = 10;
    public static final int ID_DOWNVOTE = 11;
    public static final int ID_DOWNVOTE_UNINTERESTING = 12;
    public static final int ID_DOWNVOTE_MISLEADING = 13;
    public static final int ID_DOWNVOTE_OFFENSIVE = 14;
    public static final int ID_DOWNVOTE_REPETITIVE = 15;
    public static final int ID_LIKE = 20;
    public static final int ID_SHARE = 21;
    public static final int ID_COMMENT = 22;
    public static final int ID_VISIBLE = 30;
    public static final int ID_HOVER = 31;
    public static final int ID_EXPAND = 32;
    public static final int ID_SHARE_FACEBOOK = 50;
    public static final int ID_SHARE_TWITTER = 51;
    public static final int ID_SHARE_PINTEREST = 52;
    public static final int ID_SHARE_REDDIT = 53;
    public static final int ID_SHARE_EMAIL = 54;
    public static final int ID_START = 70;
    public static final int ID_FIRST_QUARTILE = 71;
    public static final int ID_MIDPOINT = 72;
    public static final int ID_THIRD_QUARTILE = 73;
    public static final int ID_COMPLETE = 74;
    public static final int ID_MUTE = 75;
    public static final int ID_UNMUTE = 76;
    public static final int ID_PAUSE = 77;
    public static final int ID_REWIND = 78;
    public static final int ID_RESUME = 79;
    public static final int ID_FULLSCREEN = 80;
    public static final int ID_EXIT_FULLSCREEN = 81;
    public static final int ID_EXPAND_VIDEO = 82;
    public static final int ID_COLLAPSE = 83;
    public static final int ID_ACCEPT_INVITATION_LINEAR = 84;
    public static final int ID_CLOSE_LINEAR = 85;
    public static final int ID_SKIP = 86;
    public static final int ID_PROGRESS = 87;
    public static final int ID_COMMENT_REPLY = 101;
    public static final int ID_COMMENT_UPVOTE = 102;
    public static final int ID_COMMENT_DOWNVOTE = 103;
    public static final int ID_CUSTOM_01 = 104;
    public static final int ID_CUSTOM_02 = 105;
    public static final int ID_CUSTOM_03 = 106;
    public static final int ID_CUSTOM_04 = 107;
    public static final int ID_CUSTOM_05 = 108;
    public static final int ID_CUSTOM_06 = 109;
    public static final int ID_CUSTOM_07 = 110;


    // event identifier
    int id;

    // url to call to track an event
    String url;

    /**
     * Returns an event identifier
     * @return numeric event id
     */
    public int getId() {
        return id;
    }

    /**
     * Returns tracking URL of custom event
     * @return tracking url
     */
    public String getUrl() {
        return url;
    }

}
